### ISSUE TITLE
Option to not persist the value of a form control

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#1671]
Currently, with Form Builder, the Text Output control is always bound to an XML node created for that purpose.

But not all calculations should be saved with the data. Some are transient.

One approach could be to mark the associated node as excluded from the saved data. But then if you reload that data, the node won't be there, and the control will be non-relevant or empty. You could imagine creating the node automatically, but then there is the question of it's position relative to siblings.

So another way would be to make the control not depend on the presence of a node at all.  At the XForms level, an `<xf:output>` control doesn't need to be bound to a node, with the `value` attribute:

```
<xf:output value="...some expression..."/>
```

But with Form Builder, we put all calculations and MIPs on `<xf:bind>` elements in the model. This separates the data model from the view, and conveniently allows using binds as variables. So using `value` is not ideal.

It would be good if we could keep using a `<xf:bind>` element for transient output fields. An idea would be to allow `<xf:bind>` to hold a calculated value and MIPs, yet not apply to an XML node:

```
<xf:bind calculate="$foo + $bar" relevant="$qty gt 0"/>
```

In Form Builder, the form author could decide whether the field must be associated with a node or not.

See [customer thread](https://basecamp.com/1721271/projects/4504110/messages/24372457).

**Comment by ebruchez** [comment]
Let's keep this issue about having simple calculated values, and keep the worksheet discussion to #1697. We should add an option to do that. This would require:

- storing the calculation as `<xf:output value="...">`
- handling variables like `$foo` (or another way to reference control values)
  - right now, bind variables are not accessible in the view
  - see also #2909

This could also help solve issues like this [StackOverflow question](http://stackoverflow.com/questions/40397904/how-to-show-the-selected-label-instead-of-the-value-in-an-orbeon-field/40408517?noredirect=1#comment68086146_40408517) where back and forth between model and view causes weird behaviors.

**Comment by avernet** [comment]
Maybe open question: how would handle the relevance / readonlyness of such a control?

**Comment by ebruchez** [comment]
Good question. Maybe still use an instance data element and a `ref`, but not use `calculate`? It's similar to what we do with `fr:explanation`.

**Comment by avernet** [comment]
Yes, that seems like a reasonable compromise, until we support #1193.

**Comment by ebruchez** [comment]
+1 from me today. One idea, in particular, is to use a calculated value to map to a PDF template field. Having the data in the database is not desirable in this case.

2026-03-17: Was for "Orbeon Demo: Vendor Application". This form has now been modified to have non-persisted calculated values for use by the PDF template.

**Comment by ebruchez** [comment]
### Option 1

In the first message above, the first option we discuss mentions:

> One approach could be to mark the associated node as excluded it from the saved data. But then if you reload that data, the node won't be there, and the control will be non-relevant or empty.

We now have Simple Data Migration, so this would not be a problem, assuming Simple Data Migration is enabled for the form. So the only thing we would need is to be able to mark the data node as "transient". This would prevent the node from being saved, sent, etc., in all cases. If we had that, then this should work.

So this first option could work, with a reasonably simple change.

### Option 2

The second option above is to use `<xf:output value="...">`. We say:

> - handling variables like `$foo` (or another way to reference control values)
>   - right now, bind variables are not accessible in the view
>   - see also #2909

We now have mechanisms to refer to control values in the view, actions, etc., as we rewrite the variable notation to a function call. So *referring* to form control values would work.

This said:

- this wouldn't allow a calculated value to refer to another calculated value by name
- readonliness/relevance would not work until we have #1193

These would be downsides compared with Option 1.

> use an instance data element and a `ref`, but not use `calculate`

This would make Option 2 closer to Option 1, but then we might as well use `calculate`.

### Conclusion

I think that Option 1, with adding the ability to prune those nodes, would work and satisfy the requirement.

**Comment by ebruchez** [comment]
UI:

- In a "Calculated Value" field, add an option to "Omit value from data".

Backend:

- Q: Do we mark the node, or the bind, or the control?
    - We could use a custom MIP: `fr:persist` value set to `false()`
- `prepareXML()` would process `fr:persist`
- Ideally: if simple data migration is disabled, run through "super" simple data migration assuming some binds have `fr:persist`.

**Comment by ebruchez** [comment]
- [x] UI
    - [x] checkbox
    - [x] read/write to custom MIP
        - [x] add `fr:persist` to list of custom MIPs in form if missing
        - [x] add to/remove from `xf:bind`
- [x] `prepareXML()` to prune upon Save
- [x] also apply to (enabled by default for new controls)
    - [x] Explanatory text
    - [x] Button
    - [x] ~~Hidden field~~
        - probably not a good idea, might not be a good default
    - [x] ~~Calculated field~~
        - probably not a good idea, might not be a good default
- [x] change default for Explanatory Text
- [x] unit test
- [x] update demo forms
- [x] what about `send`? should that be an option?
- [x] don't disable, but select when children selection is made
- [x] doc
- [x] "Encrypt data at rest": disable/indent
- [x] "Index" also!
- [x] P2: update to Simple Data Migration

**Comment by ebruchez** [comment]
Q: Should any field have the option of not being persisted? In which case this could be a common setting to all form controls. 
A: I think it is ok to do this as a common setting.

**Comment by ebruchez** [comment]
Only issue remaining: the scenario where simple data migration is disabled. Currently, this feature won't work because the missing elements will not be recreated.

So we'd need, for `DataMigrationBehavior.Disabled` and `DataMigrationBehavior.Error`, to still run the algorithm, and then only perform the addition of missing elements which match `fr:persist="false()"`.

**Comment by ebruchez** [comment]
Following code/feature review, we need tor review the logic of the checkboxes in the UI: if a control is not persisted, then it should also not be possible to:

- index it
- encrypt it

**Comment by ebruchez** [comment]
Fix "By default":

<img width="331" height="297" alt="Image" src="https://github.com/user-attachments/assets/dedc1d5e-1adf-4432-8f0c-c25d3fd4269b" />

**Comment by ebruchez** [comment]
- [ ] confirm behavior with CE: supported? should use "simple" Simple Data Migration
